inventory.py

```diff
@@ -167,14 +180,115 @@ def check_if_valid_user_id(id: int, user_ids: Sequence[int]) -> bool:
+        return
+
+    events = audit_events['events']
+
+    for event in events:
+        if event['event_type'] == 'published':
+            course_id = event['links']['course']
+            published_date_found = True
+            published_course_date.update({course_id: event['created_at']})
+            logger.info(f"Published Date {event['created_at']} for course {course_id}")
+            break
+    if not published_date_found:
+        get_next_page_url(response)
+
+    seconds = time.time() - start_time
+    str_time = time.strftime("%H:%M:%S", time.gmtime(seconds))
+    logger.debug(f"Parsing the published date took {str_time} ")
+    return
+
+
+def get_published_course_date(course_ids, next_page_links=None):
+    logger.info("Starting of get_published_course_date call")
+    with FuturesSession(max_workers=PUBLISHED_DATE_WORKERS) as session:
+        headers = {'Content-type': 'application/json', 'Authorization': 'Bearer ' + CANVAS_TOKEN}
+        if next_page_links is not None:
+            logger.info("Going through Next page URL set")
+            responses = []
+            for next_page_link in next_page_links:
+                response = session.get(next_page_link, headers=headers)
+                published_course_next_page_list.remove(next_page_link)
+                responses.append(response)
+        else:
+            logger.info("Initial Round of Fetching course published date")
+            responses = [
+                session.get(f'{CANVAS_URL}/api/v1/audit/course/courses/{course_id}?per_page=100', headers=headers)
+                for course_id in course_ids]
+
+        for response in as_completed(responses):
+            published_date_resp_parsing(response)
+
+    if len(published_course_next_page_list) != 0:
+        logger.info(f"""Pagination size {len(published_course_next_page_list)} with published_at date items 
+                         {published_course_next_page_list}""")
+        get_published_course_date(course_ids, published_course_next_page_list)
+
+
 def run_course_inventory() -> None:
-    start = datetime.now()
+    start = time.time()
 
     # Gather course data
     course_df = gather_course_info_for_account(ACCOUNT_ID, TERM_ID)
+    course_available_df= course_df.loc[course_df.workflow_state == 'available'].copy()
+    course_available_ids = course_available_df['canvas_id'].to_list()
+    get_published_course_date(course_available_ids)
+    df1 = pd.DataFrame(published_course_date.items(), columns=['canvas_id','published_at'])
+    course_complete_df = pd.merge(course_df, df1, on='canvas_id', how='left')
+    course_complete_df['created_at'] = pd.to_datetime(course_complete_df['created_at'], format="%Y-%m-%dT%H:%M:%SZ",
+                                                      errors='coerce')
+    course_complete_df['published_at'] = pd.to_datetime(course_complete_df['published_at'], format="%Y-%m-%dT%H:%M:%SZ",
+                                                        errors='coerce')
```

Comment:
The code in Line 285-287 was done so that inserts to DB with null is only allowed when `publised_at` and `created_at` columns are DateTime instead of null. `errors='coerce'` it is enforcing `Nat` when None in those columns

Reply:
This is cool! I would like to see the `to_datetime` calls spaced out like below, but not required.

```
course_complete_df['published_at'] = pd.to_datetime(
   course_complete_df['published_at'], 
   format="%Y-%m-%dT%H:%M:%SZ",
   errors='coerce'
)
```

Reply:
Done